Initial work on SS4 upgrade

mlewis-everley opened the pull request:
So I have been working on an upgrade for this module and I thought I would submit a PR. What I have done so far:

1. Namespace classes.
2. Convert to SS4 vendor module.
3. Add support for multiple recipients and senders (as swiftmailer supports this by default).
4. General code tidy up / PSR 4 / PHPCBF.
5. Removed custom logic for members in the queue script, instead adding a more obvious `getEmailAddresses` method to members via an extension.

TODO:

1. I had to switch to storing recipients and senders as JSON in the DB (I am not sure of a better way of accommodating a list), if this is kept for the finall SS4 version final version there will need to be a migration script.

NOTES:

1. As mentioned above, I had to use json to store the sender and recipients lists. The To and From fields are now legacy
2. Probably some other stuff, but I cannot remember...

mlewis-everley committed:
Update composer to SS4 vendor module

mlewis-everley committed:
Add namespaces

mlewis-everley committed:
Add upgrade config

mlewis-everley committed:
Upgrade codebase

mlewis-everley committed:
Add autoload composer config

mlewis-everley committed:
Update inline with SS4 email class and breakout interface

mlewis-everley committed:
Tidy up namespacing

mlewis-everley committed:
inspect code

mlewis-everley committed:
Add crontask as a requirement

mlewis-everley committed:
PHPCBF

mlewis-everley committed:
Rename code to src

mlewis-everley committed:
PSR code tidy and updates

mlewis-everley committed:
Add extension to member to add email provider method

mlewis-everley committed:
Further code tidy and add support for multiple senders and recipients

Taitava commented on src/MemberExtension.php:
```diff
@@ -0,0 +1,20 @@
+ */
+class MemberExtension extends DataExtension
+{
+    public function getEmailAddresses(): array
```
I'll go through the code better later, but just a small thing I noticed now. PHPDoc could be used to denote that the return value is `string[]` (an array of strings), which is even more detailed in addition to the `: array`, which is also good.
```suggestion
    /**
     * @return string[] Email addresses as keys, member names as values.
     */
    public function getEmailAddresses(): array
```

It's been a while since I've last coded with PHP, so I don't know if it possible to define `string[]` directly as a return type like this: `public function getEmailAddresses(): string[]`?

mlewis-everley replied:
I tried adding `string[]` as the method output, but (at least for PHP7.4) this is not allowed.

Taitava commented:
Hi, thank you a lot for this pull request! 🙂 

I haven't yet tested or checked the changes. I haven't worked on my SilverStripe modules for a couple of years as I've done some changes in my work (not so much web site projects currently). I will take a better look at this PR later.


>     1. Namespace classes.
>     2. Convert to SS4 vendor module.
👍 

>     3. Add support for multiple recipients and senders (as swiftmailer supports this by default).

- [ ] Good. I just need to remember mention this in release notes.

> 
>     4. General code tidy up / PSR 4 / PHPCBF.

Really good! I'm not very good in PSR things, and PHPCBF is a new thing to me.

> 
>     5. Removed custom logic for members in the queue script, instead adding a more obvious `getEmailAddresses` method to members via an extension.

- [ ] Ok, I can't rember what the custom logic for members was, so I need to check that later.
> 
> 
> TODO:
> 
>     1. I had to switch to storing recipients and senders as JSON in the DB (I am not sure of a better way of accommodating a list), if this is kept for the finall SS4 version final version there will need to be a migration script.
> 

This sounds like a situation where we don't have good options available. If we would follow the best practises on SQL database structuring, we would add a new table for recipients, which would allow a one-to-many recipients relationship (one message can have many recipients), but I'm not so sure if I'd like that.

I'd like to avoid JSON here, if possible, it just adds so much "noise" when there are `[, `]` and `"` characters present. I think some email clients use comma `,` as a separator between email addresses? Can a comma be a legal part of an email address? I hope not. At least Gmail supports plus characters `+` in email addresses, so maybe other special characters are legal in email addresses? If comma can be used in an email address, then we cannot use it as a separator here. How about semicolon `;`?

If there's any kind of need for querying all email messages from the queue that has a certain email address as the recipient, then this kind of joining multiple email addresses together will make it harder to perform the query (need to use `LIKE '%email@address.com%'` instead of `= 'email@address.com'`). Maybe I would still consider creating the new table, although it's not a simple thing either...

So, as a summary, thank you for your effort and I will check this more deeply later and comment more! 😎

mlewis-everley commented:
I have been thinking about your comments (RE: support for multiple senders/recipients). I think to be honest, your initial suggestion of using a `has_many` association to an object with a name/email combo makes the most sense. You could create a generic `DataObject` that could be re-used for senders, recipients and in the future `CC` and `BCC`. I guess there is the issue that you will end up with potentially lots of records, but realistically all that data is going to have to be stored in some form or other in the database and a relation is a lot easier to query than JSON.

Also, I did look into just storing emails as a CSV, but swiftmailer requires senders as an array of addresses and names, EG:

```
  "address@someperson.com" => "Person One"
  "someoneelse@something.com" => "Person Two"
```
It makes sense to me at least to try and mimic this setup as best as possible?

mlewis-everley committed:
Add factory for queuing emails

mlewis-everley committed:
Use Many_Many_Through to manage To, From, CC and BCC

mlewis-everley committed:
Add some simple unit tests

mlewis-everley commented:
@Taitava I have just updated this PR with some changes:

1. I have switched `From`, `To`, `CC`, `BCC` to now be associations via a `ManyManyThrough`. This means there can be one central base table that is used for all data. As this could result in a lot of redundant data, I have also added a `findOrMake` so that emails can be linked to existing records dynamically.
2. As `EmailQueue` is now getting quite complex and logic heavy, I have created a separate factory to handle the process of queueing emails.
3. I have added a couple of unit tests, just to check some of the more complex logic.

mlewis-everley committed:
Update travis config

mlewis-everley committed:
PHPCBF

mlewis-everley committed:
Add phpcs config

mlewis-everley committed:
More travis fixes

mlewis-everley committed:
Remove php 7.2 from travis

mlewis-everley committed:
Add PHPUnit and codesniffer as dev dependencies

mlewis-everley committed:
Fix typo for postgres

mlewis-everley committed:
Change method decloration

mlewis-everley committed:
Add additional test for simple email setup

mlewis-everley committed:
Only set admin email as sender if valid

mlewis-everley committed:
Ensure that name is only included if valid

mlewis-everley committed:
PHPCBF (again apparently)

mlewis-everley commented:
Right, also I have got travis working for automated builds, which always seems to be a massive pain!

Taitava commented:
> Also, I did look into just storing emails as a CSV, but swiftmailer requires senders as an array of addresses and names, EG:
> 
> ```
>   "address@someperson.com" => "Person One"
>   "someoneelse@something.com" => "Person Two"
> ```
> 
> It makes sense to me at least to try and mimic this setup as best as possible?

Sorry, I didn't understand this. Does this mean that there always needs to be a name for each recipient, e.g. _Person One_? I'm not so convinced that providing real names would give much benefit for automated email sending. If switfmailer requires an associative array format for the recipients, can we just provide the email address twice:

```
  "address@someperson.com" => "address@someperson.com"
  "someoneelse@something.com" => "someoneelse@something.com"
```

This way we don't need to store the recipient's name in the database. Sorry if my answer was completely offtopic to what you tried to say.

> @Taitava I have just updated this PR with some changes:
> 
>     1. I have switched `From`, `To`, `CC`, `BCC` to now be associations via a `ManyManyThrough`. This means there can be one central base table that is used for all data. As this could result in a lot of redundant data, I have also added a `findOrMake` so that emails can be linked to existing records dynamically.

A many-many relation sounds good to me so that it prevents duplicate email addresses being recorded in the database. I was actually also thinking about it when I posted my previous reply, but I decided to write about has-many relation because it's simpler and easier if data needs to be queried manually (e.g. when debugging). But, many-many is good so let's use that.

I think I haven't heard of a _ManyMany**Through**_ relation in SilverStripe. Is it something that didn't exist in SS3? Or then I have just forgotten about it.



> * As `EmailQueue` is now getting quite complex and logic heavy, I have created a separate factory to handle the process of queueing emails.
> 
> * I have added a couple of unit tests, just to check some of the more complex logic.

Sounds good.

I still haven't checked these changes, though, and don't know yet when I could check them. I'm beginning to think about should I actually transfer this project to you? I mean you could have your own repository targeted for SS4, I would mark this library in packagist.org as abandoned with a mention to use your package instead. This repository would still be here to serve SS3 users (I know at least one project that still uses this library with SS3, and probably won't be upgraded to SS4). And this repository would also have a link to your repository saying _If you are using SS4, use the new repository_. What do you think about this? You can also suggest something else.
